pkg/vpcmodel/commonConnectivity.go

```diff
@@ -6,7 +6,139 @@ SPDX-License-Identifier: Apache-2.0
+		allConn:         allConn,
+	}
+}
+
+// NewConnWithStatefulGivenTCPStatefulAndNonTCP constructor that is given the (tcp stateful and non tcp) conn and the entire conn
+func NewConnWithStatefulGivenTCPStatefulAndNonTCP(tcpStatefulAndNonTCP, allConn *connection.Set) *ConnWithStateful {
+	tcpStatefulFraction, nonTCPFraction := partitionTCPNonTCP(tcpStatefulAndNonTCP)
+	return &ConnWithStateful{
+		statefulConn:    tcpStatefulFraction,
+		nonStatefulConn: computeNonStatefulConn(allConn, nonTCPFraction, tcpStatefulFraction),
+		otherConn:       nonTCPFraction,
+		allConn:         allConn,
+	}
+}
+
+func NewConnWithStatefulGivenStateful(stateful *connection.Set) *ConnWithStateful {
+	return &ConnWithStateful{
+		statefulConn:    stateful,
+		nonStatefulConn: NoConns(),
+		otherConn:       NoConns(),
+		allConn:         stateful,
+	}
+}
+
+func NewConnWithStatefulAllStateful() *ConnWithStateful {
+	return &ConnWithStateful{
+		statefulConn:    newTCPSet(),
+		nonStatefulConn: NoConns(),
+		otherConn:       NoConns(),
+		allConn:         AllConns(),
+	}
+}
+
+func NewConnWithStatefulAllNotStateful() *ConnWithStateful {
+	return &ConnWithStateful{
+		statefulConn:    NoConns(),
+		nonStatefulConn: newTCPSet(),
+		otherConn:       AllConns().Subtract(newTCPSet()),
+		allConn:         AllConns(),
+	}
+}
+
+func (e *ConnWithStateful) IsAllObliviousStateful() bool {
+	return e.allConn.Equal(connection.All())
+}
+
+func (e *ConnWithStateful) IsEmpty() bool {
+	return e.allConn.IsEmpty()
+}
+
+func (e *ConnWithStateful) Equal(other *ConnWithStateful) bool {
+	return e.statefulConn.Equal(other.statefulConn) && e.otherConn.Equal(other.otherConn) &&
+		e.allConn.Equal(other.allConn)
+}
+
+func (e *ConnWithStateful) Copy() *ConnWithStateful {
+	return NewConnWithStateful(e.nonStatefulConn.Copy(), e.otherConn.Copy(), e.allConn.Copy())
+}
+
+func (e *ConnWithStateful) Intersect(other *ConnWithStateful) *ConnWithStateful {
+	statefulConn := e.statefulConn.Intersect(other.statefulConn)
+	otherConn := e.otherConn.Intersect(other.otherConn)
+	conn := e.allConn.Intersect(other.allConn)
+	return NewConnWithStateful(statefulConn, otherConn, conn)
+}
+
+func (e *ConnWithStateful) Union(other *ConnWithStateful) *ConnWithStateful {
+	statefulConn := e.statefulConn.Union(other.statefulConn)
+	otherConn := e.otherConn.Union(other.otherConn)
+	conn := e.allConn.Union(other.allConn)
+	return NewConnWithStateful(statefulConn, otherConn, conn)
+}
+
+func (e *ConnWithStateful) Subtract(other *ConnWithStateful) *ConnWithStateful {
+	statefulConn := e.statefulConn.Subtract(other.statefulConn)
+	otherConn := e.otherConn.Subtract(other.otherConn)
+	conn := e.allConn.Subtract(other.allConn)
+	return NewConnWithStateful(statefulConn, otherConn, conn)
+}
+
+func (e *ConnWithStateful) String() string {
+	return e.allConn.String()
+}
+
+func (e *ConnWithStateful) EnhancedString() string {
+	if !e.nonStatefulConn.IsEmpty() {
+		return e.String() + " *"
+	}
+	return e.String()
+}
```

Comment:
Shouldn't rename `EnhancedString()` to be `String()`, and remove the simple `String()` method? 
If one is looking for connection string without consideration of statefuleness, he can just use `e.allConn.String()`

Reply:
No. There are cases String() is used - where we want to print the connection without the * for stateful. E.g. in explainability and in the debug report

Reply:
I still find id confusing.. shouldn't the main method `String()` include stateful information? 
and have a dedicated method such as `ConnWithoutStateful()` ?

Reply:
In 2nd though, I've deleted `string()` and used `String()` of `allConns` in the places it was used. Then I've renamed `enhancedString()` to `string()`